### ISSUE TITLE
prefer-destructuring false-positive if already declared in upper scope

### ISSUE THREAD

**Issue by JPeer264** [eslint/eslint#9734]
**Tell us about your environment**

* **ESLint Version:** 4.13.1
* **Node Version:** 8.9.1
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

airbnb-styleguide

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const foo = { bar: 1 };
let bar;

try {
  // throws an error here
  bar = foo.bar
} catch(e) {
}
```

```js
// rule setting, as of the airbnb styleguide
'prefer-destructuring': ['error', {
  VariableDeclarator: {
    array: false,
    object: true,
  },
  AssignmentExpression: {
    array: true,
    object: true,
  },
}, {
  enforceForRenamedProperties: false,
}],
```

**What did you expect to happen?**
I expect that the destructing error won't show up. As the variable `bar` is already declared in an upper scope not in the same.

**What actually happened? Please include the actual, raw output from ESLint.**
The rule suggested to prefer destructing although it won't work with destructing here, as I have to declare a new variable.

**Comment by platinumazure** [comment]
You can destructure without a declaration:

```js
const foo = { bar: 1 };
let bar;

try {
  // throws an error here
  ({ bar } = foo)
} catch(e) {
}
```

Unfortunately, you need parentheses for object destructuring due to the language grammar, but otherwise destructuring does work without a declaration.

**Comment by JPeer264** [comment]
Ah alright, didn't thought the parentheses do the trick. Thanks!

**Comment by mrchief** [comment]
Parenthesis solution gets rid of the error but throws a wrench at readability. Linter should be smart enough here to not throw the error .

**Comment by mrchief** [comment]
Just saw this 3fc4e3f7a5e0d04950490ed8af08dfba0b270ecc. So maybe the linter _**is**_ getting smarter! :)

**Comment by mrchief** [comment]
@platinumazure Thanks for taking the time.

I get the parenthesis part. However, what I read from the OP, that the linter should know that a variable is declared in outer scope, so it should not throw the error in the first place. Now in case of the code sample in the OP, it seems that declaration and assignment can be combined (assuming try block do anything with the variables) but there could be cases like these:

```
let fields = []

if(someCondition) {
  fields = grabFromSomewhere(...)
} else {
  fields = grabbedFromSomewhereElse.fields
}
```

In other words, the rule should warn when declaring stuff, not otherwise. (I realize that makes it look like I'm asking for another option to the rule, not sure if we already have one).

I also realize I can rewrite the else part to use parenthesis but that just looks odd (and one may not immediately grasp why the parenthesis are there).

**Comment by mrchief** [comment]
I also want to point out that I'm aware of `AssignmentExpression` and `VariableDeclarator` options. What I meant with "another option" is something like this:

should not warn for cases like these
```
let fields = []

if(someCondition) {
  fields = grabFromSomewhere(...)
} else {
  fields = grabbedFromSomewhereElse.fields
}
``` 

should warn for cases like these 
```
let fields = []

if(someCondition) {
  fields = grabbedFromSomewhereElse.fields
}
``` 
because above can be potentially refactored to use destructuring. 

Not a great example but I hope you understand. And probably I'm asking for a lot! :)
